crates/flipperzero/src/furi/stream_buffer.rs

```diff
@@ -0,0 +1,500 @@
+//! Furi stream buffer primitive.
+
+use core::{
+    ffi::c_void,
+    fmt::{Debug, Display},
+    num::NonZeroUsize,
+    ptr::NonNull,
+};
+
+use crate::furi;
+
+use flipperzero_sys::{self as sys, furi::Status};
+use ufmt::{derive::uDebug, uDebug, uDisplay, uwrite};
+
+/// Furi stream buffer primitive.
+///
+/// Stream buffers are used to send a continous stream of data from one task or interrupt to another.
+/// Their implementation is light weight, making them particularly suited for interrupt to task and
+/// core to core communication scenarios.
+///
+/// # Safety
+///
+/// Stream buffer implementation assumes there is only one task or interrupt that will write to the
+/// buffer (the writer), and only one task or interrupt that will read from the buffer (the reader).
+/// This behavior has to be carefully ensured when using [`send`](Self::send) and
+/// [`receive`](Self::receive) directly.
+///
+/// For safer usage, consider the
+#[cfg_attr(not(feature = "alloc"), doc = "`Sender`")]
+#[cfg_attr(feature = "alloc", doc = "[`Sender`]")]
+/// and
+#[cfg_attr(not(feature = "alloc"), doc = "`Receiver`,")]
+#[cfg_attr(feature = "alloc", doc = "[`Receiver`],")]
+/// available via the `alloc` feature, to send and receive data.
+#[derive(Debug)]
+pub struct StreamBuffer(NonNull<sys::FuriStreamBuffer>);
+
+impl uDebug for StreamBuffer {
+    fn fmt<W>(&self, f: &mut ufmt::Formatter<'_, W>) -> Result<(), W::Error>
+    where
+        W: ufmt::uWrite + ?Sized,
+    {
+        f.debug_tuple("StreamBuffer")?
+            .field(&self.0.as_ptr())?
+            .finish()
+    }
+}
+
+// SAFETY:
+// The Furi API does not impose any restrictions on moving a stream buffer between threads.
+// Since the API permits this, `StreamBuffer` can safely implement `Send`.
+unsafe impl Send for StreamBuffer {}
+
+// SAFETY:
+// The Furi API requires that there be only one writer and one reader at any given time.
+// However, both the writer and reader may be moved between threads.
+// This ensures that using the stream buffer across threads is safe, provided that the
+// one-writer-one-reader rule is upheld.
+// The responsibility for maintaining safety while sending and receiving data lies within the
+// `send` and `receive` methods.
+unsafe impl Sync for StreamBuffer {}
+
+impl StreamBuffer {
+    /// Create a new instance of a `StreamBuffer`.
+    ///
+    /// The `trigger_level` defines the number of bytes that must be present in the stream buffer
+    /// before any blocked tasks waiting for data can proceed.
+    ///
+    /// For sending and receiving data safely use
+    #[cfg_attr(not(feature = "alloc"), doc = "`into_stream`")]
+    #[cfg_attr(feature = "alloc", doc = "[`into_stream`](Self::into_stream)")]
+    /// which is available using the `alloc` feature.
+    pub fn new(size: NonZeroUsize, trigger_level: usize) -> Self {
+        let size: usize = size.into();
```

Comment:
Non-blocker nit (ptobably, for the future changes): given that type annotation is already used, `from` seems more appropriate:

```suggestion
        let size = usize::from(size);
```